src/caffe/layers/bn_layer.cpp

```diff
@@ -0,0 +1,349 @@
+        Dtype(1), spatial_mean_.cpu_diff(),
+        batch_sum_multiplier_.cpu_data(),
+        Dtype(0), shift_diff);
+
+// Propagate down
+
+    // put scale * top_diff to buffer_blob_
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_, C_, 1, Dtype(1),
+        batch_sum_multiplier_.cpu_data(), scale_data, Dtype(0),
+        spatial_variance_.mutable_cpu_data());
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_ * C_,
+        H_ * W_, 1, Dtype(1),
+        spatial_variance_.cpu_data(),
+        spatial_sum_multiplier_.cpu_data(), Dtype(0),
+        buffer_blob_.mutable_cpu_data());
+    caffe_mul(buffer_blob_.count(), top_diff, buffer_blob_.cpu_data(),
+        buffer_blob_.mutable_cpu_data());
+
+    // use new top diff for computation
+    caffe_mul(buffer_blob_.count(),  x_norm_.cpu_data(),
+        buffer_blob_.cpu_data(), bottom_diff);
+    // EX across spatial
+    caffe_cpu_gemv<Dtype>(CblasNoTrans, N_ * C_, H_ * W_,
+        Dtype(1), bottom_diff,
+        spatial_sum_multiplier_.cpu_data(), Dtype(0),
+        spatial_mean_.mutable_cpu_data());
+    // EX across batch
+    caffe_cpu_gemv<Dtype>(CblasTrans, N_, C_, Dtype(1),
+        spatial_mean_.cpu_data(),
+        batch_sum_multiplier_.cpu_data(), Dtype(0),
+        batch_mean_.mutable_cpu_data());
+
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans,
+        N_, C_, 1, Dtype(1),
+        batch_sum_multiplier_.cpu_data(),
+        batch_mean_.cpu_data(), Dtype(0),
+        spatial_mean_.mutable_cpu_data());
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans, N_ * C_,
+        H_ * W_, 1, Dtype(1),
+        spatial_mean_.cpu_data(),
+        spatial_sum_multiplier_.cpu_data(), Dtype(0),
+        bottom_diff);
+
+    caffe_mul(buffer_blob_.count(),
+        x_norm_.cpu_data(), bottom_diff, bottom_diff);
+
+    // EX across spatial
+    caffe_cpu_gemv<Dtype>(CblasNoTrans, N_ * C_,
+        H_ * W_, Dtype(1), buffer_blob_.cpu_data(),
+        spatial_sum_multiplier_.cpu_data(), Dtype(0),
+        spatial_mean_.mutable_cpu_data());
+    // EX across batch
+    caffe_cpu_gemv<Dtype>(CblasTrans, N_, C_, Dtype(1),
+        spatial_mean_.cpu_data(),
+        batch_sum_multiplier_.cpu_data(), Dtype(0),
+        batch_mean_.mutable_cpu_data());
+
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans,
+        N_, C_, 1, Dtype(1),
+        batch_sum_multiplier_.cpu_data(),
+        batch_mean_.cpu_data(), Dtype(0),
+        spatial_mean_.mutable_cpu_data());
+    caffe_cpu_gemm<Dtype>(CblasNoTrans, CblasNoTrans,
+        N_ * C_, H_ * W_, 1, Dtype(1),
+        spatial_mean_.cpu_data(),
+        spatial_sum_multiplier_.cpu_data(), Dtype(1), bottom_diff);
+
+    caffe_cpu_axpby(buffer_blob_.count(), Dtype(1),
+        buffer_blob_.cpu_data(), Dtype(-1. / (N_ * H_ * W_)),
+        bottom_diff);
+
+    // put the squares of bottom into buffer_blob_
+    caffe_powx(buffer_blob_.count(), bottom_data, Dtype(2),
+        buffer_blob_.mutable_cpu_data());
```

Comment:
line 326 - 327 is redundant and can be deleted. So is the same for the GPU code.